src/backend/distributed/utils/resource_lock.c

```diff
@@ -760,7 +765,88 @@ lock_relation_if_exists(PG_FUNCTION_ARGS)
 	relation = makeRangeVarFromNameList(relationNameList);
 
     /* lock the relation with the lock mode */
-	RangeVarGetRelid(relation, lockMode, false);
+	RangeVarGetRelidInternal(relation, lockMode, 0, CitusRangeVarCallbackForLockTable,
+							 (void *) &lockMode);
 
 	PG_RETURN_BOOL(true);
 }
+
+
+/*
+ * CitusRangeVarCallbackForLockTable is a callback for RangeVarGetRelidExtended used
+ * to check whether the user has permission to lock a table in a particular mode.
+ *
+ * This function is a copy of RangeVarCallbackForLockTable in lockcmds.c adapted to
+ * Citus code style.
+ */
+static void
+CitusRangeVarCallbackForLockTable(const RangeVar *rangeVar, Oid relationId,
+								  Oid oldRelationId, void *arg)
+{
+	LOCKMODE lockmode = *(LOCKMODE *) arg;
+	char relkind;
+	AclResult aclResult;
+
+	if (!OidIsValid(relationId))
+	{
+        /* table doesn't exist, so no permissions check */
+		return;
+	}
+
+	relkind = get_rel_relkind(relationId);
+	if (!relkind)
+	{
+        /* table was concurrently dropped */
+		return;
+	}
+
+    /* we only allow tables and views to be locked */
+	if (relkind != RELKIND_RELATION && relkind != RELKIND_PARTITIONED_TABLE &&
```

Comment:
We have `RegularTable(Oid)` which could be useful for getting rid of pg 96 warnings for the first two.